contract/models/contract_line.py

```diff
@@ -0,0 +1,819 @@
+                # in post-paid case.
+                # i.e: date_end = 15/03 recurring_next_date = 31/03
+                # A solution for this, is to only check recurring_next_date
+                # and filter with create_invoice_visibility
+            ]
+        )
+        lines = self.search(domain).filtered('create_invoice_visibility')
+        if lines:
+            return lines._recurring_create_invoice()
+        return False
+
+    @api.multi
+    def _recurring_create_invoice(self):
+        """Create invoices from contracts
+
+        :return: invoices created
+        """
+        invoices = self.env['account.invoice']
+        for contract in self.mapped('contract_id'):
+            lines = self.filtered(lambda l: l.contract_id == contract)
+            invoices |= lines._create_invoice()
+            lines._update_recurring_next_date()
+        return invoices
+
+    @api.multi
+    def _create_invoice(self):
+        """
+        :param invoice: If not False add lines to this invoice
+        :return: invoice created or updated
+        """
+        contract = self.mapped('contract_id')
+        date_invoice = min(self.mapped('recurring_next_date'))
+        invoice = self.env['account.invoice'].create(
+            contract._prepare_invoice(date_invoice)
+        )
+        for line in self:
+            invoice_line_vals = line._prepare_invoice_line(invoice.id)
+            if invoice_line_vals:
+                self.env['account.invoice.line'].create(invoice_line_vals)
+        invoice.compute_taxes()
+        return invoice
+
+    @api.multi
+    def _prepare_invoice_line(self, invoice_id):
+        self.ensure_one()
+        invoice_line = self.env['account.invoice.line'].new(
+            {
+                'invoice_id': invoice_id,
+                'product_id': self.product_id.id,
+                'quantity': self.quantity,
+                'uom_id': self.uom_id.id,
+                'discount': self.discount,
+            }
+        )
+        # Get other invoice line values from product onchange
+        invoice_line._onchange_product_id()
+        invoice_line_vals = invoice_line._convert_to_write(invoice_line._cache)
+        # Insert markers
+        contract = self.contract_id
+        lang_obj = self.env['res.lang']
+        lang = lang_obj.search([('code', '=', contract.partner_id.lang)])
+        date_format = lang.date_format or '%m/%d/%Y'
+        name = self._insert_markers(date_format)
+        invoice_line_vals.update(
+            {
+                'name': name,
+                'account_analytic_id': contract.id,
+                'price_unit': self.price_unit,
+            }
+        )
+        return invoice_line_vals
+
+    @api.multi
+    def _insert_markers(self, date_format):
+        self.ensure_one()
+        date_from = fields.Date.from_string(self.recurring_next_date)
+        date_to = date_from + self.get_relative_delta(
+            self.recurring_rule_type, self.recurring_interval
+        )
```

Comment:
This is not a good place to compute date_from and date_to, they should be passed as parameter to this function, since they must have been computed elsewhere when deciding what we are invoicing.

The logic that computes the period being invoiced deserves it's how method, with unit tests.